# Basics (simple )Lineare Regression

## Quellen und Links

### Implementierung der OLS mit SK-Learn !Wertvoll!
[LinkNotebookImplementationUD-Course--USA-HOUSING](./Literatur_Quellen/1_UDC/original/Python-DataScience-MachineLearning-v10/5-Machine%20Learning/1-Linear%20Regression/1-Lineare%20Regression%20Mit%20Python.ipynb) das ist eine BSP-Regressions-Implementierung für ein Regressionsproblem aus dem Rene-Brunner Kurs. Ganz klassisch ein Housing-Data-Problem

### Implementierung der OLS mit SK-Learn !Wertvoll!
[LinkNotebookImplementationUD-Course--Customer-Analysis](./Literatur_Quellen/1_UDC/original/Python-DataScience-MachineLearning-v10/5-Machine%20Learning/1-Linear%20Regression/1-Lineare%20Regression%20Mit%20Python.ipynb) das ist eine BSP-Regressions-Implementierung für ein Regressionsproblem aus dem Rene-Brunner Kurs. Ganz klassisch eine Kunden-Analyse

<hr>

### Implementierung der OLS mit SK-Learn !Quick & Dirty = keine Kommentare!
[Boston-Housing](./Literatur_Quellen/1_UDC/original/Python-DataScience-MachineLearning-v10/5-Machine%20Learning/1-Linear%20Regression/1-Lineare%20Regression%20Mit%20Python.ipynb) Nach dem Vorbild des Oberen USA-Housing

<hr>


### Vergleich zwischen "Statsmodel" und  "SKLearn" for Linear Regression
https://becominghuman.ai/stats-models-vs-sklearn-for-linear-regression-f19df95ad99b unter diesem Link werden zwei unterschiedliche Module verglichen mit welchen die Lineare Regression durchgeführt werden kann


## Algemeines Vorgehen SK-Learn
1. [EDA](https://github.com/JHC90/Basic-DataScience-Skills/wiki/EDA_Landingpage)<br>
Aus der EDA lese ich unter anderem ab, was im Data-Preprocessing bzw Feature Engineering zu tun ist.
2. [Data-Preprocessing]()
3. [Modellierung Regression (SK-Learn), Predicten, MSE, MAE, RMSE](./Literatur_Quellen/1_UDC/original/Python-DataScience-MachineLearning-v10/5-Machine%20Learning/1-Linear%20Regression/1-Lineare%20Regression%20Mit%20Python.ipynb)


#### R²

In [29]:
print('Coefficient of determination: %.2f'
      % r2_score(y, y_pred_SKlearn))

Coefficient of determination: 0.93


### [statsmodel](https://www.statsmodels.org/devel/generated/statsmodels.regression.linear_model.OLS.html)
hat den vorteil, dass eine art Summary-Function inkludiert ist. 

#### Modeling

In [30]:
import statsmodels.api as sm
X = sm.add_constant(X)
modelStatsmodel = sm.OLS(y, X).fit()

#### Summary = R², adjR²,P-Value, F-test,T-Test,AIC,BIC, ModelCoefficentsen

In [23]:
modelStatsmodel.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.926
Model:                            OLS   Adj. R-squared:                  0.926
Method:                 Least Squares   F-statistic:                     1841.
Date:                Fri, 20 Mar 2020   Prob (F-statistic):           5.03e-85
Time:                        08:33:15   Log-Likelihood:                 23.742
No. Observations:                 149   AIC:                            -43.48
Df Residuals:                     147   BIC:                            -37.48
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.3661      0.040     -9.064      0.000      -0.446      -0.286
x1             0.4163      0.010     42.905      0.000       0.397       0.436
==============================================================================
Omnibus:                        5.326   Durbin-Watson:                   1.461
Prob(Omnibus):                  0.070   Jarque-Bera (JB):                4.993
Skew:                           0.352   Prob(JB):                       0.0824
Kurtosis:                       3.556   Cond. No.                         10.4
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

wenn wir nun die Koeffizienten betrachten, sehen wir, dass diese die gleichen sind mit const = -0.3661 && x1 = 0.4163 wie beider Methode von SKLEARN. AUßerdem ist auch der r² der selbe wie bei SKLEARN

#### Predcition

In [31]:
y_pred_StatsModel = modelStatsmodel.predict(X)

#### MSE & RMSE
Da habe ich bisher auf die schnelle von statsmodel noch nichts passendes gefunden => daher verwende ich die SK-Learn Methode

In [43]:
print('Mean squared error: %.2f'
      % mean_squared_error(y, y_pred_StatsModel))

Mean squared error: 0.04


### Zusammenfassung Compare Stats Models vs SKLearn for Linear Regression

## Interpreatition der entsprechenden Werte
innerhalb der Regression fallen folgende Werte an. Dazu habe ich im Github-Repo ein Wiki erstellt, in welchem ich die Werte erkläre und dort Update. <br>
Als Datengrundlage verwende ich die Summary von statsmodel aus dem vorherigen Beispiel des IRIS-Datensatzes

In [47]:
modelStatsmodel.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.926
Model:                            OLS   Adj. R-squared:                  0.926
Method:                 Least Squares   F-statistic:                     1841.
Date:                Fri, 20 Mar 2020   Prob (F-statistic):           5.03e-85
Time:                        08:57:17   Log-Likelihood:                 23.742
No. Observations:                 149   AIC:                            -43.48
Df Residuals:                     147   BIC:                            -37.48
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.3661      0.040     -9.064      0.000      -0.446      -0.286
x1             0.4163      0.010     42.905      0.000       0.397       0.436
==============================================================================
Omnibus:                        5.326   Durbin-Watson:                   1.461
Prob(Omnibus):                  0.070   Jarque-Bera (JB):                4.993
Skew:                           0.352   Prob(JB):                       0.0824
Kurtosis:                       3.556   Cond. No.                         10.4
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### [MSE & RMSE](https://github.com/JHC90/Basic-DataScience-Skills/wiki/Lineare-Regression_4_MeasureQuality_MSE_RMSE)

### [P-Value](https://github.com/JHC90/Basic-DataScience-Skills/wiki/Lineare-Regression_5_P-Value)

### [R² -Value](https://github.com/JHC90/Basic-DataScience-Skills/wiki/R%C2%B2-Value)

### [adj-R² -Value](https://github.com/JHC90/Basic-DataScience-Skills/wiki/Lineare-Regression_7_Adjusted-R%C2%B2)

### [t-statistic](https://github.com/JHC90/Basic-DataScience-Skills/wiki/Lineare-Regression_9_t-Statistics)

### [F-statistic](https://github.com/JHC90/Basic-DataScience-Skills/wiki/Lineare-Regression_8_F-Statistics)

### [AIC ](https://github.com/JHC90/Basic-DataScience-Skills/wiki/Lineare-Regression_10_AIC)

### [BIC ](https://github.com/JHC90/Basic-DataScience-Skills/wiki/Lineare-Regression_11_BIC)